src/GraphQL/Resolver.hs

```diff
@@ -134,17 +133,24 @@ data Result a = Result [ResolverError] a deriving (Show, Functor, Eq)
 aggregateResults :: [Result Value] -> Result Value
 aggregateResults r = toValue <$> sequenceA r
 
+throwE :: Applicative f => ResolverError -> f (Result Value)
+throwE err = pure (Result [err] GValue.ValueNull)
+
 instance Applicative Result where
   pure v = Result [] v
   (Result e1 f) <*> (Result e2 x) = Result (e1 <> e2) (f x)
 
 ok :: Value -> Result Value
 ok = pure
 
+type AllTypes = Map Name TypeDefinition
 
 class HasResolver m a where
   type Handler m a
-  resolve :: Handler m a -> SelectionSet Value -> m (Result Value)
+  -- XXX: Might be nicer to make AllTypes a ReaderT -- easier to ignore when
+  -- not wanted, and wouldn't have required me to change so many call sites.
+  -- OTOH, with all our type weirdness, I got bored trying to figure it out.
+  resolve :: AllTypes -> Handler m a -> Maybe (SelectionSetByType Value) -> m (Result Value)
```

Comment:
All interesting points!

The API (as envisioned by my so far) would be (in the simple case) that I have a new type, e.g.

```haskell
data BusinessLogicThing = MoneyYes { amount :: Int } deriving Generic
```

and then 

```haskell
instance HasResolver BusinessLogicThing where
   resolve handler _ = map businessLogicThingToValue handler
```

The more complex (and probably more interesting) case is that users implement new combinators, e.g. `ReadonlyObject` or `DynamicSchemaObject` (the latter could e.g. figure out what fields it supports on the fly). 

I think at this stage I agree with `AllTypes` being a stand-in for the full schema, and think we should keep it.

I'm also capitulating in terms of design, I really need some larger-project feedback now before making any other decisions.

Summary: :shipit: